SingularityService/src/main/java/com/hubspot/singularity/scheduler/SingularityNewTaskChecker.java

```diff
@@ -370,21 +381,50 @@ CheckTaskState getTaskState(SingularityTask task, Optional<SingularityRequestWit
     return Optional.absent();
   }
 
-  private boolean isOverdue(SingularityTask task) {
+  private boolean isHealthcheckOverdue(SingularityTask task) {
+    final long healthcheckDuration = System.currentTimeMillis() - getTaskRunningStartTime(task.getTaskId());
+
+    final boolean isOverdue = healthcheckDuration > getKillAfterHealthcheckRunningForMillis();
+
+    if (isOverdue) {
+      LOG.debug("Task {} healthcheck is overdue (duration: {}), allowed limit {}", task.getTaskId(), JavaUtils.durationFromMillis(healthcheckDuration), JavaUtils.durationFromMillis(getKillAfterHealthcheckRunningForMillis()));
+    }
+
+    return isOverdue;
+  }
+
+  private boolean isTaskOverdue(SingularityTask task) {
     final long taskDuration = System.currentTimeMillis() - task.getTaskId().getStartedAt();
 
-    final boolean isOverdue = taskDuration > getKillAfterUnhealthyMillis();
+    final boolean isOverdue = taskDuration > getKillAfterTaskNotRunningMillis();
 
     if (isOverdue) {
-      LOG.debug("Task {} is overdue (duration: {}), allowed limit {}", task.getTaskId(), JavaUtils.durationFromMillis(taskDuration), JavaUtils.durationFromMillis(getKillAfterUnhealthyMillis()));
+      LOG.debug("Task {} is overdue (duration: {}), allowed limit {}", task.getTaskId(), JavaUtils.durationFromMillis(taskDuration), JavaUtils.durationFromMillis(getKillAfterTaskNotRunningMillis()));
     }
 
     return isOverdue;
   }
 
+  private long getTaskRunningStartTime(SingularityTaskId task) {
+    Optional<SingularityTaskHistory> taskHistory = taskManager.getTaskHistory(task);
+
+    if (taskHistory.isPresent()) {
+      java.util.Optional<SingularityTaskHistoryUpdate> taskRunningState = taskHistory.get().getTaskUpdates().stream().filter(h -> h.getTaskState().equals(ExtendedTaskState.TASK_RUNNING)).findFirst();
+
+      if (taskRunningState.isPresent()) {
+        return taskRunningState.get().getTimestamp();
+      }
+
+      LOG.error("Could not find time when task {} reached TASK_RUNNING state", task);
+    } else {
+      LOG.error("Could not find task history for {}", task);
+    }
+
+    return System.currentTimeMillis();
+  }
```

Comment:
had it fall back to the current time if it doesn't find the state info. could run over in this case, but we should be notified by one of those error messages